File: app/src/main/java/com/sample/lmn/davide/cachefilesample/MainActivity.java
package com.sample.lmn.davide.cachefilesample;

import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.sample.lmn.davide.cachefilesample.components.DaggerDownloadManagerComponent;
import com.sample.lmn.davide.cachefilesample.manager.CacheManager;
import com.sample.lmn.davide.cachefilesample.manager.DownloadSoundtrackManager;
import com.sample.lmn.davide.cachefilesample.modules.DownloadManagerModule;

import java.lang.ref.WeakReference;

import javax.inject.Inject;

public class MainActivity extends AppCompatActivity implements View.OnClickListener,
        CacheManager.OnCacheEntryRetrievesCallbacks,
        Response.Listener, Response.ErrorListener {
    private static final String FILENAME_SAMPLE = "mozart_sample.mp3";
    private static final String REMOTE_FILE = "http://www.amclassical.com/mp3/amclassical_moonlight_sonata_movement_1.mp3";
    private View putOnCacheButton;
    private View getOnCacheButton;
    private View requestFileButton;

    @Inject
    public CacheManager cacheManager;
    @Inject
    public DownloadSoundtrackManager downloadSoundtrackManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //building dagger component
        DaggerDownloadManagerComponent
                .builder()
                .downloadManagerModule(new DownloadManagerModule(getApplicationContext(), this))
                .build()
                .inject(this);

        //init view
        onInitView();
    }

    private void onInitView() {
        putOnCacheButton = findViewById(R.id.putOnCacheButtonId);
        getOnCacheButton = findViewById(R.id.getOnCacheButtonId);
        requestFileButton = findViewById(R.id.requestFileButtonId);
        putOnCacheButton.setOnClickListener(this);
        getOnCacheButton.setOnClickListener(this);
        requestFileButton.setOnClickListener(this);
    }


    @Override
    public void onClick(final View view) {
        switch (view.getId()) {
            case R.id.putOnCacheButtonId:
                cacheManager.put(FILENAME_SAMPLE);
                break;
            case R.id.getOnCacheButtonId:
                cacheManager.getAsync(REMOTE_FILE);
                break;
            case R.id.requestFileButtonId:
                downloadSoundtrackManager
                        .getFileFromUrl(Uri.parse(REMOTE_FILE));
                downloadSoundtrackManager.setLst(new WeakReference<Response.Listener<byte[]>>(this));
                downloadSoundtrackManager.setLst2(new WeakReference<Response.ErrorListener>(this));
                break;
        }
    }

    private void handleResult(Object file, Context context) {
        if (file instanceof byte[]) {
            if (((byte[]) file).length != 0) {
                Toast.makeText(context, "file size" + ((byte[]) file).length, Toast.LENGTH_SHORT).show();
                return;
            }

            Toast.makeText(context, "Empty file", Toast.LENGTH_SHORT).show();
            return;
        }


        if (file != null &&
                file.toString().length() != 0) {
            Toast.makeText(context, file.toString(), Toast.LENGTH_SHORT).show();
            return;
        }

        Toast.makeText(context, "Empty file", Toast.LENGTH_SHORT).show();

    }

    @Override
    public void onCacheEntryRetrieved(final Object file) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                handleResult(file, getApplicationContext());
            }
        });
    }

    @Override
    public void onErrorResponse(VolleyError error) {
        Log.e(getClass().getName(), "download error - " + error.getMessage());

    }

    @Override
    public void onResponse(Object response) {
        Toast.makeText(this, "Download with success " + ((byte[]) response).length, Toast.LENGTH_SHORT).show();
        Log.e(getClass().getName(), "download ok- " + ((byte[]) response).length);
    }

}

File: app/src/main/java/com/sample/lmn/davide/cachefilesample/MainActivity.kt
package com.sample.lmn.davide.cachefilesample

import android.media.AudioManager
import android.media.MediaPlayer
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import com.google.api.services.youtube.model.SearchResult
import com.lib.lmn.davide.soundtrackdownloaderlibrary.manager.SoundTrackDownloaderManager
import com.lib.lmn.davide.soundtrackdownloaderlibrary.modules.SoundTrackDownloaderModule
import com.sample.lmn.davide.cachefilesample.adapter.OnSoundTrackItemClickListener
import com.sample.lmn.davide.cachefilesample.adapter.SoundTrackRvAdapter
import com.sample.lmn.davide.cachefilesample.managers.YoutubeOnSearchByQueryResults
import com.sample.lmn.davide.cachefilesample.managers.YoutubeV3AuthenticatorManager
import kotlinx.android.synthetic.main.activity_main.*
import java.io.FileInputStream
import java.lang.ref.WeakReference

open class MainActivity : AppCompatActivity() , SoundTrackDownloaderModule.OnSoundTrackRetrievesCallbacks,
        YoutubeOnSearchByQueryResults {

    lateinit var mediaPlayer: MediaPlayer

    val soundTrackDownloaderManager: SoundTrackDownloaderManager by lazy {
        SoundTrackDownloaderManager.getInstance(this, this)
    }

    val youtubeDownloadManager: YoutubeV3AuthenticatorManager by lazy {
        YoutubeV3AuthenticatorManager("bla", WeakReference(this))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        //init view
        recyclerViewId.layoutManager = LinearLayoutManager(applicationContext)
        recyclerViewId.adapter = SoundTrackRvAdapter(ArrayList(), listener = object: OnSoundTrackItemClickListener {
            override fun onItemClick(item: SearchResult) {
                soundTrackDownloaderManager.downloadAndPlaySoundTrack(item.id.videoId)
            }
        })
        onInitView()
    }

    /**
     *
     */
    private fun onInitView() {
        clearButtonId.setOnClickListener { soundTrackUrlEditTextId.setText("") }
        playButtonId.setOnClickListener {
            youtubeDownloadManager.searchByQuery(soundTrackUrlEditTextId.text.toString())
        }

        initMediaPlayer()
    }

    /**
     *
     */
    private fun initMediaPlayer() {
        mediaPlayer = MediaPlayer()
        mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC)
    }

    /**
     *
     */
    private fun playCachedFile(inputStream: FileInputStream) {
        try {
            showSuccess("eureka")
            mediaPlayer.reset()
            mediaPlayer.setDataSource(inputStream.fd)
            mediaPlayer.prepare()
            mediaPlayer.start()
        } catch (e: Exception) {
            e.printStackTrace()
            showError(e.message?: "default")
        }
    }

    /**
     *
     */
    override fun onSoundTrackRetrieveSuccess(fileInputStream: FileInputStream) {
        runOnUiThread {
            playCachedFile(fileInputStream)
            showSuccess("sound track successfully downloaded")
        }
    }

    /**
     *
     */
    override fun onSoundTrackRetrieveError(message: String?) {
        showError("error" + message)
    }

    /**
     *
     */
    override fun youtubeSearchResultSuccess(list: List<SearchResult>, query: String?) {
        showSuccess("$query --- ${list.size}")
        (recyclerViewId.adapter as SoundTrackRvAdapter).list = list
        recyclerViewId.adapter.notifyDataSetChanged()
    }

    /**
     *
     */
    override fun youtubeSearchResultError(message: String, query: String?) {
        showError("error" + message)
    }

    /**
     *
     */
    fun showSuccess(message: String) {
        Snackbar.make(findViewById(R.id.mainActivityLayoutId), message, Snackbar.LENGTH_LONG).show()
    }

    /**
     *
     */
    fun showError(message: String) {
        Snackbar.make(findViewById(R.id.mainActivityLayoutId), message, Snackbar.LENGTH_LONG).show()
    }

    /**
     *
     */
    companion object {
        private val FILENAME_SAMPLE = "mozart_sample.mp3"
        private val REMOTE_FILE = "http://www.amclassical.com/mp3/amclassical_moonlight_sonata_movement_1.mp3"
    }

}
